File: guilib-bukkit/src/main/java/net/developertobi/guilib/bukkit/api/item/GuiItemImpl.java
package net.developertobi.guilib.bukkit.api.item;

import net.developertobi.guilib.api.GuiItemAction;
import net.developertobi.guilib.api.gui.GuiController;
import net.developertobi.guilib.api.item.GuiItem;
import net.developertobi.guilib.api.item.GuiPos;
import net.developertobi.guilib.api.item.ItemEnchantment;
import net.developertobi.guilib.api.utils.MathUtils;
import net.kyori.adventure.text.Component;
import org.bukkit.Material;
import org.bukkit.event.inventory.InventoryClickEvent;
import org.bukkit.inventory.Inventory;
import org.bukkit.inventory.ItemStack;

import java.time.Instant;
import java.util.List;
import java.util.UUID;

public class GuiItemImpl implements GuiItem {

    private ItemStack item;
    private final GuiItemAction action;
    private final String itemId = UUID.randomUUID().toString().split("-")[0];
    private Instant lastClickTime = Instant.EPOCH;
    private final long cooldownMillis = 250;

    public GuiItemImpl(ItemStack item) {
        this(item, null);
    }

    public GuiItemImpl(ItemStack item, GuiItemAction action) {
        this.item = item;
        this.action = action;
    }

    @Override
    public ItemStack getItem() {
        return item;
    }

    @Override
    public void setItem(ItemStack item) {
        this.item = item;
    }

    String getItemId() {
        return itemId;
    }

    @Override
    public void runAction(GuiPos position, GuiItem guiItem, InventoryClickEvent event) {
        Instant now = Instant.now();

        if (!lastClickTime.equals(Instant.EPOCH) && (now.toEpochMilli() - lastClickTime.toEpochMilli()) < cooldownMillis) {
            event.setCancelled(true);
            return;
        }

        lastClickTime = now;
        if (action != null) {
            action.run(position, guiItem, event);
        }
    }

    @Override
    public void update(GuiController controller, GuiItem.Modification modification, Object... values) {
        if (values.length > 1) {
            throw new UnsupportedOperationException("There are no more than one value allowed! Current size: " + values.length);
        }

        GuiPos guiPosition = controller.getPositionOfItem(this);
        ItemStack modifiableItem;
        ItemStack extraItem = null;

        if (guiPosition != null) {
            int slot = MathUtils.positionToSlot(guiPosition, controller.getColumns());
            Inventory rawInventory = controller.getRawInventory();
            if (rawInventory == null) {
                throw new IllegalStateException("Raw inventory is null");
            }
            modifiableItem = rawInventory.getItem(slot);
            if (modifiableItem == null) {
                modifiableItem = this.item;
            }
        } else {
            modifiableItem = this.item;
        }

        if (controller.getPagination() != null) {
            for (GuiItem paginationItem : controller.getPagination().getItems().values()) {
                if (paginationItem instanceof GuiItemImpl impl && impl.getItemId().equals(this.itemId)) {
                    extraItem = impl.getItem();
                    break;
                }
            }
        }

        Object newValue = values[0];

        switch (modification) {
            case TYPE -> {
                if (!(newValue instanceof Material material)) {
                    throw new UnsupportedOperationException("'newValue' is not a Material!");
                }

                modifiableItem = modifiableItem.withType(material);
                this.item = modifiableItem;

                if (guiPosition != null) {
                    int slot = MathUtils.positionToSlot(guiPosition, controller.getColumns());
                    Inventory rawInventory = controller.getRawInventory();
                    if (rawInventory != null) {
                        rawInventory.setItem(slot, modifiableItem);
                    }
                }

                if (extraItem != null) {
                    extraItem = extraItem.withType(material);
                }
                if (extraItem != null && controller.getPagination() != null) {
                    for (GuiItem paginationItem : controller.getPagination().getItems().values()) {
                        if (paginationItem instanceof GuiItemImpl impl && impl.getItemId().equals(this.itemId)) {
                            impl.setItem(extraItem);
                            break;
                        }
                    }
                }
            }

            case DISPLAY_NAME -> {
                if (!(newValue instanceof Component component)) {
                    throw new UnsupportedOperationException("'newValue' is not a Component!");
                }

                modifiableItem.editMeta(meta -> meta.displayName(component));
                if (extraItem != null) {
                    extraItem.editMeta(meta -> meta.displayName(component));
                }
            }

            case LORE -> {
                if (!(newValue instanceof List<?> list)) {
                    throw new UnsupportedOperationException("'newValue' is not a List!");
                }

                @SuppressWarnings("unchecked")
                List<Component> lore = (List<Component>) list;
                modifiableItem.editMeta(meta -> meta.lore(lore));
                if (extraItem != null) {
                    extraItem.editMeta(meta -> meta.lore(lore));
                }
            }

            case AMOUNT -> {
                if (!(newValue instanceof Integer amount)) {
                    throw new UnsupportedOperationException("'newValue' is not an Integer!");
                }

                modifiableItem.setAmount(amount);
                if (extraItem != null) {
                    extraItem.setAmount(amount);
                }
            }

            case INCREMENT -> {
                if (!(newValue instanceof Integer increment)) {
                    throw new UnsupportedOperationException("'newValue' is not an Integer!");
                }

                modifiableItem.setAmount(modifiableItem.getAmount() + increment);
                if (extraItem != null) {
                    extraItem.setAmount(extraItem.getAmount() + increment);
                }
            }

            case ENCHANTMENTS -> {
                if (!(newValue instanceof ItemEnchantment enchantment)) {
                    throw new UnsupportedOperationException("'newValue' is not an ItemEnchantment!");
                }

                enchantment.performAction(modifiableItem);
                if (extraItem != null) {
                    enchantment.performAction(extraItem);
                }
            }

            case GLOWING -> {
                if (!(newValue instanceof Boolean glowing)) {
                    throw new UnsupportedOperationException("'newValue' is not a Boolean!");
                }

                modifiableItem.editMeta(meta -> meta.setEnchantmentGlintOverride(glowing));
                if (extraItem != null) {
                    extraItem.editMeta(meta -> meta.setEnchantmentGlintOverride(glowing));
                }
            }
        }
    }
}

File: guilib-bukkit/bin/main/net/developertobi/inventorylib/bukkit/api/item/GuiItemImpl.kt
package net.developertobi.inventorylib.bukkit.api.item

import net.kyori.adventure.text.Component
import net.developertobi.inventorylib.api.inventory.GuiController
import net.developertobi.inventorylib.api.item.GuiItem
import net.developertobi.inventorylib.api.item.GuiPos
import net.developertobi.inventorylib.api.item.ItemEnchantment
import net.developertobi.inventorylib.api.utils.MathUtils
import org.bukkit.Material
import org.bukkit.event.inventory.InventoryClickEvent
import org.bukkit.inventory.Inventory
import org.bukkit.inventory.ItemStack
import org.bukkit.inventory.meta.ItemMeta
import java.time.Instant
import java.util.*

class GuiItemImpl(
    override var item: ItemStack,
    private val action: ((GuiPos, GuiItem, InventoryClickEvent) -> Unit)?
) : GuiItem {

    private val itemId: String = UUID.randomUUID().toString().split("-")[0]
    private var lastClickTime: Instant = Instant.EPOCH
    private var cooldownMillis: Long = 250

    override fun runAction(position: GuiPos, guiItem: GuiItem, event: InventoryClickEvent) {
        val now = Instant.now()

        if (lastClickTime != Instant.EPOCH && (now.toEpochMilli() - lastClickTime.toEpochMilli()) < cooldownMillis) {
            event.isCancelled = true
            return
        }

        lastClickTime = now
        action?.invoke(position, guiItem, event)
    }

    @Suppress("UNCHECKED_CAST")
    override fun update(controller: GuiController, modification: GuiItem.Modification, vararg values: Any) {
        if (values.size > 1) throw UnsupportedOperationException("There are no more than one value allowed! Current size: " + values.size)

        val inventoryPosition: GuiPos? = controller.getPositionOfItem(this)
        var modifiableItem: ItemStack
        var extraItem: ItemStack? = null

        if (inventoryPosition != null) {
            val slot: Int = MathUtils.positionToSlot(inventoryPosition, controller.getColumns())
            val rawInventory: Inventory = controller.rawInventory!!
            modifiableItem = rawInventory.getItem(slot) ?: this.item
        } else {
            modifiableItem = this.item
        }

        if (controller.pagination != null)
            extraItem = controller.pagination!!.items.values().find { it is GuiItemImpl && it.itemId == this.itemId }?.item

        val newValue = values[0]

        when (modification) {
            GuiItem.Modification.TYPE -> {
                if (newValue !is Material) throw UnsupportedOperationException("'newValue' is not a Material!")

                modifiableItem = modifiableItem.withType(newValue)
                this.item = modifiableItem

                if (inventoryPosition != null) {
                    val slot: Int = MathUtils.positionToSlot(inventoryPosition, controller.getColumns())
                    controller.rawInventory!!.setItem(slot, modifiableItem)
                }

                extraItem = extraItem?.withType(newValue)
                if (extraItem != null) {
                    val paginationItem = controller.pagination?.items?.values()?.find { it is GuiItemImpl && it.itemId == this.itemId } as? GuiItemImpl
                    paginationItem?.item = extraItem
                }
            }

            GuiItem.Modification.DISPLAY_NAME -> {
                if (newValue !is Component) throw UnsupportedOperationException("'newValue' is not a Component!")

                modifiableItem.editMeta { itemMeta: ItemMeta -> itemMeta.displayName(newValue) }
                extraItem?.editMeta { itemMeta: ItemMeta -> itemMeta.displayName(newValue) }
            }

            GuiItem.Modification.LORE -> {
                if (newValue !is MutableList<*>) throw UnsupportedOperationException("'newValue' is not a List!")

                modifiableItem.editMeta { it.lore(newValue as MutableList<Component>) }
                extraItem?.editMeta { it.lore(newValue as MutableList<Component>) }
            }

            GuiItem.Modification.AMOUNT -> {
                if (newValue !is Int) throw UnsupportedOperationException("'newValue' is not an Integer!")

                modifiableItem.amount = newValue
                if (extraItem != null) extraItem.amount = newValue
            }

            GuiItem.Modification.INCREMENT -> {
                if (newValue !is Int) throw UnsupportedOperationException("'newValue' is not an Integer!")

                modifiableItem.amount += newValue
                if (extraItem != null) extraItem.amount += newValue
            }

            GuiItem.Modification.ENCHANTMENTS -> {
                if (newValue !is ItemEnchantment) throw UnsupportedOperationException("'newValue' is not an ItemEnchantment!")

                newValue.performAction(modifiableItem)
                if (extraItem != null) newValue.performAction(extraItem)
            }

            GuiItem.Modification.GLOWING -> {
                if (newValue !is Boolean) throw UnsupportedOperationException("'newValue' is not an Boolean!")

                modifiableItem.editMeta {
                    it.setEnchantmentGlintOverride(newValue)
                }
                extraItem?.editMeta {
                    it.setEnchantmentGlintOverride(newValue)
                }
            }
        }
    }
}


